3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/singleton.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2012-2017 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at
+ * http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and
+ * limitations there under.
+ */
+package com.snowplowanalytics
+package snowplow.enrich
+package spark
+
+// Jackson
+import com.fasterxml.jackson.databind.JsonNode
+
+// Snowplow
+import common.{FatalEtlError, ValidatedMessage, ValidatedNelMessage}
+import common.utils.{ConversionUtils, JsonUtils}
+import iglu.client.Resolver
+import iglu.client.validation.ProcessingMessageMethods._
+
+/** Singletons needed for unserializable classes. */
+object singleton {
+  /** Singleton for Iglu's Resolver to maintain one Resolver per node. */
+  object ResolverSingleton {
+    @volatile private var instance: Resolver = _
+    /**
+     * Retrieve or build an instance of Iglu's Resolver.
+     * @param igluConfig JSON representing the Iglu configuration
+     */
+    def get(igluConfig: String): Resolver = {
+      if (instance == null) {
+        synchronized {
+          if (instance == null) {
+            instance = getIgluResolver(igluConfig)
+              .valueOr(e => throw new FatalEtlError(e.toString))
+          }
+        }
+      }
+      instance
+    }
+
+    /**
+     * Build an Iglu resolver from a JSON.
+     * @param json JSON representing the Iglu resolver
+     * @return A Resolver or one or more error messages boxed in a Scalaz ValidationNel
+     */
+    private[spark] def getIgluResolver(json: String): ValidatedNelMessage[Resolver] =
+      for {
+        node <- (base64ToJsonNode(json, "iglu").toValidationNel: ValidatedNelMessage[JsonNode])
+        reso <- Resolver.parse(node)
+      } yield reso
+  }
+
+  /** Singleton for EnrichmentRegistry. */
+  object RegistrySingleton {
+    import common.enrichments.EnrichmentRegistry
+    @volatile private var instance: EnrichmentRegistry = _
+    @volatile private var enrichments: String = _
+    /**
+     * Retrieve or build an instance of EnrichmentRegistry.
+     * @param igluConfig JSON representing the Iglu configuration
+     * @param enrichments JSON representing the enrichments that need performing
+     * @param local Whether to build a registry from local data
+     */
+    def get(igluConfig: String, enrichments: String, local: Boolean): EnrichmentRegistry = {
+      if (instance == null || this.enrichments != enrichments) {
+        synchronized {
```

Comment:
Just out of curiosity - is `synchronized` really necessary? I mean, are these objects really accessed in multi-threading env?